src/payment/strategies/googlepay/googlepay-payment-strategy.ts

```diff
@@ -65,38 +74,39 @@ export default class GooglePayPaymentStrategy implements PaymentStrategy {
             .then(() => this._store.getState());
     }
 
-    execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+    async execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
         if (!this._googlePayOptions) {
             throw new InvalidArgumentError('Unable to initialize payment because "options.googlepay" argument is not provided.');
         }
 
-        const {
-            onError = () => {},
-            onPaymentSelect = () => {},
-        } = this._googlePayOptions;
-
-        return Promise.resolve(this._getPayment())
-            .then(payment => {
-                if (!payment.paymentData.nonce || !payment.paymentData.cardInformation) {
-                    // TODO: Find a way to share the code with _handleWalletButtonClick method
-                    return this._googlePayPaymentProcessor.displayWallet()
-                        .then(paymentData => this._paymentInstrumentSelected(paymentData))
-                        .then(() => onPaymentSelect())
-                        .then(() => this._getPayment())
-                        .catch(error => {
-                            if (error.statusCode !== 'CANCELED') {
-                                onError(error);
-                            }
-                        });
-                }
-
-                return payment;
-            })
-            .then(() =>
-                this._store.dispatch(this._orderActionCreator.submitOrder({ useStoreCredit: payload.useStoreCredit }, options))
-                    .then(() => this._store.dispatch(this._paymentActionCreator.submitPayment(this._getPayment())))
-                    .catch(error => this._googlePayAdyenV2PaymentProcessor?.processAdditionalAction(error) || Promise.reject(error))
-            );
+        if (!payload.payment) {
+            throw new PaymentArgumentInvalidError(['payment']);
+        }
+
+        const { methodId } = payload.payment;
+
+        let payment = await this._getPayment(methodId);
+
+        if (!payment.paymentData.nonce || !payment.paymentData.cardInformation) {
+            const {
+                onError = () => {},
+                onPaymentSelect = () => {},
+            } = this._googlePayOptions;
+            await this._displayWallet(methodId, onPaymentSelect, onError);
+            payment = await this._getPayment(methodId);
+        }
+
+        if (!payment.paymentData.nonce) {
+            throw new MissingDataError(MissingDataErrorType.MissingPayment);
+        }
+
+        try {
+            await this._store.dispatch(this._orderActionCreator.submitOrder({ useStoreCredit: payload.useStoreCredit }, options));
+
+            return await this._store.dispatch(this._paymentActionCreator.submitPayment(payment));
+        } catch (error) {
+            return this._googlePayAdyenV2PaymentProcessor?.processAdditionalAction(error) || Promise.reject(error);
```

Comment:
rather than doing `||`, I think what you actually want is to execute both
```js
catch(error) {
   this._googlePayAdyenV2PaymentProcessor?.processAdditionalAction(error);
   throw error;
}
```

re-throwing the error should behave the same as `Promise.reject(error)` but simpler.